control/noise/handshake.go

```
@@ -0,0 +1,430 @@
+	// generation code clamps keys to avoid low order points. I
+	// believe that makes pub equivalent to the output of
+	// X25519(privateKey, Basepoint), and so the contract is
+	// respected.
+	keyData, err := curve25519.X25519(priv[:], pub[:])
+	if err != nil {
+		return fmt.Errorf("computing X25519: %w", err)
+	}
+
+	r := hkdf.New(newBLAKE2s, keyData, s.ck[:], nil)
+	if _, err := io.ReadFull(r, s.ck[:]); err != nil {
+		return fmt.Errorf("extracting ck: %w", err)
+	}
+	if _, err := io.ReadFull(r, s.k[:]); err != nil {
+		return fmt.Errorf("extracting k: %w", err)
+	}
+	s.n = 0
+	return nil
+}
+
+// EncryptAndHash encrypts plaintext into ciphertext (which must be
+// the correct size to hold the encrypted plaintext) using the current
+// s.k, mixes the ciphertext into s.h, and returns the ciphertext.
+func (s *symmetricState) EncryptAndHash(ciphertext, plaintext []byte) {
+	if s.n == invalidNonce {
+		// Noise in general permits writing "ciphertext" without a
+		// key, but in IK it cannot happen.
+		panic("attempted encryption with uninitialized key")
+	}
+	if len(ciphertext) != len(plaintext)+poly1305.TagSize {
+		panic("ciphertext is wrong size for given plaintext")
+	}
+	aead := newCHP(s.k)
+	var nonce [chp.NonceSize]byte
+	binary.BigEndian.PutUint64(nonce[4:], s.n)
+	s.n++
+	ret := aead.Seal(ciphertext[:0], nonce[:], plaintext, s.h[:])
+	s.MixHash(ret)
+}
+
+// DecryptAndHash decrypts the given ciphertext into plaintext (which
+// must be the correct size to hold the decrypted ciphertext) using
+// the current s.k. If decryption is successful, it mixes the
+// ciphertext into s.h.
+func (s *symmetricState) DecryptAndHash(plaintext, ciphertext []byte) error {
+	if s.n == invalidNonce {
+		// Noise in general permits "ciphertext" without a key, but in
+		// IK it cannot happen.
+		panic("attempted encryption with uninitialized key")
+	}
+	if len(ciphertext) != len(plaintext)+poly1305.TagSize {
+		panic("plaintext is wrong size for given ciphertext")
```

I know this is impossible when using the message types, but I am a little worried about DecryptAndHash being later used on attacker-controlled buffers, leading to the panic. In general, when you can return an error, I learned to do that instead and save myself a Denial of Service. :)